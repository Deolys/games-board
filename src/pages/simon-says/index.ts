export { SimonSaysGame } from "./simon-says";