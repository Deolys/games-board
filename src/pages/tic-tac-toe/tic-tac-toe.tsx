import { useState } from 'react';
import { Box, Button, List, ListItem, Stack, Typography } from "@mui/material";
import { Header } from '../../components/header';

interface SquareProps {
  value: 'X' | 'O' | null;
  onSquareClick: () => void;
}

interface BoardProps {
  xIsNext: boolean;
  squares: Array<'X' | 'O' | null>;
  onPlay: (squares: Array<'X' | 'O' | null>) => void;
}

function Square({ value, onSquareClick }: SquareProps) {
  return (
    <Button variant="outlined" sx={{ width: 70, height: 70, m: 0.5 }} onClick={onSquareClick}>
      {value}
    </Button>
  );
}

function Board({ xIsNext, squares, onPlay }: BoardProps) {
  function handleClick(i: number) {
    if (calculateWinner(squares) || squares[i]) {
      return;
    }
    const nextSquares = squares.slice();
    if (xIsNext) {
      nextSquares[i] = 'X';
    } else {
      nextSquares[i] = 'O';
    }
    onPlay(nextSquares);
  }

  const winner = calculateWinner(squares);
  let status: string;
  if (winner) {
    status = 'Winner: ' + winner;
  } else {
    status = 'Next player: ' + (xIsNext ? 'X' : 'O');
  }

  if (!squares.includes(null)  &&  winner ===  null ){
    status = 'It\'s a tie';
  }

  return (
    <>
      <Header />
      <Typography sx={{textAlign: "center", my: 2}}>{status}</Typography>
      <Stack sx={{flexDirection: "row"}}>
        <Square value={squares[0]} onSquareClick={() => handleClick(0)} />
        <Square value={squares[1]} onSquareClick={() => handleClick(1)} />
        <Square value={squares[2]} onSquareClick={() => handleClick(2)} />
      </Stack>
      <Stack sx={{flexDirection: "row"}}>
        <Square value={squares[3]} onSquareClick={() => handleClick(3)} />
        <Square value={squares[4]} onSquareClick={() => handleClick(4)} />
        <Square value={squares[5]} onSquareClick={() => handleClick(5)} />
      </Stack>
      <Stack sx={{flexDirection: "row"}}>
        <Square value={squares[6]} onSquareClick={() => handleClick(6)} />
        <Square value={squares[7]} onSquareClick={() => handleClick(7)} />
        <Square value={squares[8]} onSquareClick={() => handleClick(8)} />
      </Stack>
    </>
  );
}

export default function TicTacToe() {
  const [history, setHistory] = useState<Array<Array<'X' | 'O' | null>>>([Array(9).fill(null)]);
  const [currentMove, setCurrentMove] = useState<number>(0);
  const xIsNext = currentMove % 2 === 0;
  const currentSquares = history[currentMove];

  const handlePlay = (nextSquares: Array<'X' | 'O' | null>) => {
    const nextHistory = history.slice(0, currentMove + 1).concat([nextSquares]);
    setHistory(nextHistory);
    setCurrentMove(nextHistory.length - 1);
  };

  const jumpTo = (nextMove: number) => {
    setCurrentMove(nextMove);
  };


  const moves = history.map((_squares, move) => {
    let description;
    if (move > 0) {
      description = 'Go to move #' + move;
    } else {
      description = 'Go to game start';
    }
    return (
      <ListItem divider key={move} sx={{display:  "flex", justifyContent: "center"}}>
        <Button variant="outlined" onClick={() => jumpTo(move)}>{description}</Button>
      </ListItem>
    );
  });

  return (
    <Box component="section" sx={{display:  "flex", flexDirection:  "column", alignItems:  "center"}}>
      <Box width="100">
        <Board xIsNext={xIsNext} squares={currentSquares} onPlay={handlePlay} />
      </Box>
      <Box>
        <List>{moves}</List>
      </Box>
    </Box>
  );
}

function calculateWinner(squares: Array<'X' | 'O' | null>): 'X' | 'O' | null {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];
  for (let i = 0; i < lines.length; i++) {
    const [a, b, c] = lines[i];
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
  }
  return null;
}