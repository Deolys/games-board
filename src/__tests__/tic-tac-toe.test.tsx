import { fireEvent, render, screen } from "@testing-library/react"
import { TicTacToe } from "../pages/tic-tac-toe"

describe('TicTacToe', () => {
  it('TicTacToe mounts properly', () => {
    const { container } = render(<TicTacToe />);
    expect(container).toMatchSnapshot();
  })

  it('When clicked, a cross or zero is placed', () => {
    const { container } = render(<TicTacToe />);
    const boardCells = container.querySelectorAll('.MuiButton-outlinedPrimary');

    fireEvent.click(boardCells[0]);
    expect(boardCells[0].innerHTML[0]).toBe('X');

    fireEvent.click(boardCells[1]);
    expect(boardCells[1].innerHTML[0]).toBe('O');
  })

  it('Claculates winner correctly', () => {
    const { container } = render(<TicTacToe />);
    const boardCells = container.querySelectorAll('.MuiButton-outlinedPrimary');
    const status = screen.getByText(/Next player: X/i);
    
    fireEvent.click(boardCells[0]);
    fireEvent.click(boardCells[4]);
    fireEvent.click(boardCells[1]);
    fireEvent.click(boardCells[5]);
    fireEvent.click(boardCells[2]);

    expect(status.innerHTML).toBe('Winner: X');
  })
})