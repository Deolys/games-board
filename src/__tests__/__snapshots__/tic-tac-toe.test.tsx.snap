// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicTacToe TicTacToe mounts properly 1`] = `
<div>
  <section
    class="MuiBox-root css-dvxtzn"
  >
    <div
      class="MuiBox-root css-1vm3mq1"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1413kmy-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-u4fbkj-MuiToolbar-root"
        >
          <a
            class="css-3qnz2e"
            href="/"
          >
            <img
              alt="Games board"
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-1waifce-MuiTypography-root"
            >
              GamesÂ board
            </p>
          </a>
          <div
            class="MuiStack-root css-1dzrg60-MuiStack-root"
          >
            <a
              class="MuiTypography-root MuiTypography-body1 css-1v5m6us-MuiTypography-root"
              href="/"
            >
              HOME
            </a>
            <button
              aria-haspopup="true"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-17ovnxk-MuiButtonBase-root-MuiButton-root"
              id="composition-button"
              tabindex="0"
              type="button"
            >
              Games
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </header>
      <p
        class="MuiTypography-root MuiTypography-body1 css-18dcqoe-MuiTypography-root"
      >
        Next player: X
      </p>
      <div
        class="MuiStack-root css-ku2emo-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiStack-root css-ku2emo-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiStack-root css-ku2emo-MuiStack-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1c3yzdo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-divider css-v53xh8-MuiListItem-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Go to game start
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </li>
      </ul>
    </div>
  </section>
</div>
`;
